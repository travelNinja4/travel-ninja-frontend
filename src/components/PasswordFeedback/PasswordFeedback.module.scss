/**
------------------------------------------------------
styles specific to the PasswordFeedback component
------------------------------------------------------
**/

.container {
  list-style: none;
  padding: 0;
  margin: 0.5rem 0 0 0;
  font-size: 0.85rem;
  line-height: 1.4;

  .subContainer {
    display: flex;
    flex-direction: row;
    align-items: center;

    li {
      position: relative;
      color: var(--color-gray-600);
    }

    .valid {
      margin-left: 0.5rem;
    }
  }
}
