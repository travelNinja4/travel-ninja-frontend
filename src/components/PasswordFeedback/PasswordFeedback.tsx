/**
 * The PasswordFeedback component provides real-time visual feedback for password requirements.
 *
 * @example
 * ```tsx
 * import PasswordFeedback from '@src/components/PasswordFeedback'
 *
 * export default function PasswordFeedback() {
 *   return<PasswordFeedback value={password} hints={passwordHints} />;
 * }
 * ```
 */

import Typography from '../Typography';
import { Dot, Check } from 'lucide-react';
import styles from './PasswordFeedback.module.scss';

/**
 * Define the props available for the PasswordFeedback component.
 */
interface PasswordFeedbackProps {
  /**
   * The current password value entered by the user.
   * Used to test against the provided password hints.
   */
  value: string;

  /**
   * An array of password requirements (hints).
   * Each hint includes a `label` describing the rule and a `regex` to validate against.
   */
  hints: PasswordHint[];
}

export interface PasswordHint {
  label: string;
  regex: RegExp;
}

export default function PasswordFeedback({ value, hints }: PasswordFeedbackProps) {
  const hasValue = value.length > 0;
  return (
    <ul data-testid="PasswordFeedbackTest" className={styles.container} aria-live="polite">
      {hints.map((hint, idx) => {
        const valid = hint.regex.test(value || '');
        return (
          <div key={idx} className={styles.subContainer}>
            {valid ? (
              <Check
                data-testid={`check-icon-${idx}`}
                color="var(--color-green-500)"
                size={20}
                className={styles.valid}
              />
            ) : (
              <Dot
                data-testid={`dot-icon-${idx}`}
                color={hasValue ? 'var(--color-red-500)' : 'var(--color-gray-600)'}
                size={30}
              />
            )}
            <Typography tag="li">{hint.label}</Typography>
          </div>
        );
      })}
    </ul>
  );
}
