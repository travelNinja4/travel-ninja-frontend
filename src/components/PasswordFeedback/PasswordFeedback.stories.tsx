import type { Meta, StoryObj } from '@storybook/nextjs-vite';
import React, { useState } from 'react';
import TextField from '../TextField';
/**
import { fn } from 'storybook/test';
*/
import PasswordFeedback from './PasswordFeedback';

const meta = {
  title: 'components/PasswordFeedback',
  component: PasswordFeedback,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    value: {
      control: 'text',
      description: 'Current password value for validation',
    },
    hints: {
      control: 'object',
      description: 'Array of password validation hints',
    },
  },
  args: {
    /** onClick: fn() **/
  },
} satisfies Meta<typeof PasswordFeedback>;

export default meta;
type Story = StoryObj<typeof meta>;

const exampleHints = [
  { label: 'At least 8 characters', regex: /.{8,}/ },
  { label: 'Contains a number', regex: /\d/ },
  { label: 'Contains a capital letter', regex: /[A-Z]/ },
  { label: 'Contains a special character', regex: /[!@#$%^&*]/ },
];

export const Interactive: Story = {
  render: (args) => {
    const [password, setPassword] = useState('');

    return (
      <div style={{ width: 300 }}>
        <TextField
          type="password"
          value={password}
          placeholder="Enter password"
          onChange={(e) => setPassword(e.target.value)}
        />
        <PasswordFeedback value={password} hints={args.hints || []} />
      </div>
    );
  },
  args: {
    value: '',
    hints: exampleHints,
  },
};
