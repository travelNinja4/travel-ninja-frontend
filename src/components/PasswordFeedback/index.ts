import PasswordFeedback from './PasswordFeedback'

export default PasswordFeedback