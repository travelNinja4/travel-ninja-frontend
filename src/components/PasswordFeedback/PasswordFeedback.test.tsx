import '@testing-library/jest-dom';
import { render, screen } from '@testing-library/react';
import PasswordFeedback from './PasswordFeedback';

const hints = [
  { label: 'At least 8 characters', regex: /.{8,}/ },
  { label: 'Contains a number', regex: /\d/ },
];

describe('Home', () => {
  /** Base props for PasswordFeedback **/
  const baseProps = {
    value: 'Test123!',
    hints: [
      { label: 'At least 8 characters', regex: /.{8,}/ },
      { label: 'Contains a number', regex: /\d/ },
      { label: 'Contains an uppercase letter', regex: /[A-Z]/ },
      { label: 'Contains a special character', regex: /[!@#$%^&*]/ },
    ],
  };

  /**
   * @test
   * Renders the PasswordFeedback component without throwing an error
   * and checks if the PasswordFeedbackTest text is present in the document.
   *
   * This ensures that the component loads correctly with the base props.
   */
  it('renders without crashing', () => {
    render(<PasswordFeedback {...baseProps} />);
    expect(screen.getByTestId('PasswordFeedbackTest')).toBeInTheDocument();
  });

  /**
   * @test
   * Ensures that the PasswordFeedback component renders without throwing any runtime errors.
   *
   * This is a safety check to confirm that the component can mount successfully
   * with the provided base props, even if it does not render any visible content.
   */
  it('renders without throwing', () => {
    expect(() => render(<PasswordFeedback {...baseProps} />)).not.toThrow();
  });

  /**
   * @test
   * Captures a snapshot of the PasswordFeedback component with base props
   * and ensures it matches the previously stored snapshot.
   *
   * This helps detect any unexpected UI changes.
   */
  it('matches snapshot', () => {
    const { asFragment } = render(<PasswordFeedback {...baseProps} />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('renders valid hints correctly', () => {
    render(<PasswordFeedback value="Password1" hints={hints} />);

    const firstHint = screen.getByText('At least 8 characters');
    const secondHint = screen.getByText('Contains a number');

    expect(firstHint).toHaveClass('valid');
    expect(secondHint).toHaveClass('valid');
  });

  it('renders invalid hints correctly', () => {
    render(<PasswordFeedback value="short" hints={hints} />);

    const firstHint = screen.getByText('At least 8 characters');
    const secondHint = screen.getByText('Contains a number');

    expect(firstHint).toHaveClass('invalid');
    expect(secondHint).toHaveClass('invalid');
  });
});
