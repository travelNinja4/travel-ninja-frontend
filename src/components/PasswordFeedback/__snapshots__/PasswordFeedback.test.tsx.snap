// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`Home matches snapshot 1`] = `
<DocumentFragment>
  <ul
    aria-live="polite"
    class="passwordHints"
    data-testid="PasswordFeedbackTest"
  >
    <li
      class="typography li align-left valid"
      data-testid="TypographyTest"
    >
      At least 8 characters
    </li>
    <li
      class="typography li align-left valid"
      data-testid="TypographyTest"
    >
      Contains a number
    </li>
    <li
      class="typography li align-left valid"
      data-testid="TypographyTest"
    >
      Contains an uppercase letter
    </li>
    <li
      class="typography li align-left valid"
      data-testid="TypographyTest"
    >
      Contains a special character
    </li>
  </ul>
</DocumentFragment>
`;
